File: java/traversals/dfs/IterativeDFS.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

public class IterativeDFS {
    // In-order traversal using iterative DFS
    public void inorder(TreeNode root) {
        if (root == null) return;

        Stack<TreeNode> stack = new Stack<>();
        TreeNode curr = root;

        while (curr != null || !stack.isEmpty()) {
            while (curr != null) {
                stack.push(curr);
                curr = curr.left;
            }

            curr = stack.pop();
            System.out.println(curr.val);
            curr = curr.right;
        }
    }

    // Pre-order traversal using iterative DFS
    public void preorder(TreeNode root) {
        if (root == null) return;

        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);

        while (!stack.isEmpty()) {
            TreeNode curr = stack.pop();
            System.out.println(curr.val);

            if (curr.right != null) {
                stack.push(curr.right);
            }
            if (curr.left != null) {
                stack.push(curr.left);
            }
        }
    }

    // Post-order traversal using iterative DFS
    public void postorder(TreeNode root) {
        if (root == null) return;

        Stack<TreeNode> stack = new Stack<>();
        Stack<Boolean> visit = new Stack<>();
        stack.push(root);
        visit.push(false);

        while (!stack.isEmpty()) {
            TreeNode curr = stack.pop();
            boolean visited = visit.pop();

            if (curr != null) {
                if (visited) {
                    System.out.println(curr.val);
                } else {
                    stack.push(curr);
                    visit.push(true);
                    stack.push(curr.right);
                    visit.push(false);
                    stack.push(curr.left);
                    visit.push(false);
                }
            }
        }
    }

    public static void main(String[] args) {
        // Create a sample binary tree
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);

        BinaryTree tree = new BinaryTree();

        System.out.println("In-order traversal:");
        tree.inorder(root);

        System.out.println("\nPre-order traversal:");
        tree.preorder(root);

        System.out.println("\nPost-order traversal:");
        tree.postorder(root);
    }
}

File: cpp/traversals/dfs/iterativeDFS.cpp
#include <iostream>
#include <stack>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class IterativeDFS {
public:
    // In-order traversal using iterative DFS
    void inorder(TreeNode* root) {
        if (root == nullptr) return;

        stack<TreeNode*> stack;
        TreeNode* curr = root;

        while (curr != nullptr || !stack.empty()) {
            while (curr != nullptr) {
                stack.push(curr);
                curr = curr->left;
            }

            curr = stack.top();
            stack.pop();
            cout << curr->val << endl;
            curr = curr->right;
        }
    }

    // Pre-order traversal using iterative DFS
    void preorder(TreeNode* root) {
        if (root == nullptr) return;

        stack<TreeNode*> stack;
        stack.push(root);

        while (!stack.empty()) {
            TreeNode* curr = stack.top();
            stack.pop();
            cout << curr->val << endl;

            if (curr->right != nullptr) {
                stack.push(curr->right);
            }
            if (curr->left != nullptr) {
                stack.push(curr->left);
            }
        }
    }

    // Post-order traversal using iterative DFS
    void postorder(TreeNode* root) {
        if (root == nullptr) return;

        stack<TreeNode*> stack;
        stack.push(root);
        stack.push(root);

        while (!stack.empty()) {
            TreeNode* curr = stack.top();
            stack.pop();

            if (!stack.empty() && curr == stack.top()) {
                if (curr->right != nullptr) {
                    stack.push(curr->right);
                    stack.push(curr->right);
                }
                if (curr->left != nullptr) {
                    stack.push(curr->left);
                    stack.push(curr->left);
                }
            } else {
                cout << curr->val << endl;
            }
        }
    }
};

int main() {
    // Create a sample binary tree
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);

    BinaryTree tree;

    cout << "In-order traversal:" << endl;
    tree.inorder(root);

    cout << "\nPre-order traversal:" << endl;
    tree.preorder(root);

    cout << "\nPost-order traversal:" << endl;
    tree.postorder(root);

    return 0;
}

